### PR TITLE
Redesign 4: Move rebase functionality into GitBranch

### PR DESCRIPTION
**Pull request by jpichon**

### CODE REVIEW

File: git_wrapper/branch.py [hunk 1]
```diff
@@ -1,51 +1,58 @@
...
+            :param repo.GitRepo git_repo: An already constructed GitRepo object to use
+            :param logging.Logger logger: A pre-configured Python Logger object
+        """
+        self.git_repo = git_repo
+        self.logger = logger
 
     @reference_exists("branch_name")
     @reference_exists("hash_")
-    def to_hash(self, branch_name, hash_):
+    def rebase_to_hash(self, branch_name, hash_):
```
Comment (jguiditta) [inline git_wrapper/branch.py]: I like this naming, very clear

File: git_wrapper/branch.py [hunk 2]
```diff
@@ -1,51 +1,58 @@
...
         """Perform a rebase from a specific reference to another.
 
            :param str branch_name: The name of the branch to rebase on
            :param str hash_: The commit hash or reference to rebase to
         """
-        self.logger.debug("Rebasing branch %s to hash %s. Repo currently at commit %s.", branch_name, hash_, self.repo.head.commit)
+        self.logger.debug("Rebasing branch %s to hash %s. Repo currently at commit %s.", branch_name, hash_, self.git_repo.repo.head.commit)
```
Comment (jguiditta) [inline git_wrapper/branch.py]: Any chance we could wrap this line?

File: git_wrapper/branch.py [hunk 3]
```diff
@@ -1,51 +1,58 @@
...
 
-        if self.repo.is_dirty():
-            msg = "Repository {0} is dirty. Please clean workspace before proceeding.".format(self.repo.working_dir)
+        if self.git_repo.repo.is_dirty():
+            msg = "Repository {0} is dirty. Please clean workspace before proceeding.".format(self.git_repo.repo.working_dir)
```
Comment (jguiditta) [inline git_wrapper/branch.py]: Same as above, wrap if you are willing

File: git_wrapper/repo.py [hunk 4]
```diff
@@ -167,3 +169,18 @@ def remote(self, new_remote):
             :param git_wrapper.remote.GitRemote new_remote: An already constructed GitRemote object to use
         """
         self._remote = new_remote
+
+    @property
+    def branch(self):
+        """Return object to act on the repo's branches"""
+        if not self._branch:
+            self._branch = GitBranch(git_repo=self, logger=self.logger)
```
Comment (jguiditta) [inline git_wrapper/repo.py]: Minor thought, and I could be wrong here - if we are calling self._branch, is the setter ever actually called?

Comment (jpichon) [reply git_wrapper/repo.py]: I guess the setter is relevant for external callers. The setter used to be called here in a previous patchset, I don't really mind either way.

Comment (jguiditta) [reply git_wrapper/repo.py]: k,it can be addressed later, if needed

**Pull request merged**